android/app/build.gradle

```diff
@@ -84,11 +84,12 @@ def osName = System.properties['os.name'].toLowerCase()
 project.ext.appId = 'app.organicmaps'
 project.ext.appName = 'Organic Maps'
 
-java {
-  toolchain {
-    languageVersion.set(JavaLanguageVersion.of(17))
-  }
-}
+// I have Java 21 installed, but this doesn't work on MacOS.
+//java {
+//  toolchain {
+//    languageVersion.set(JavaLanguageVersion.of(17))
```

Comment:
This piece looks like needed/required to have stable/reproducible builds everywhere: https://developer.android.com/build/jdks#toolchain

1. Does it fail only in Android Studio or also from command line?
2. Did you try to kill/restart gradle daemon? `./gradlew --stop` may help
3. Did you try to clean your android/.idea folder and reopen the project in Android Studio?

Reply:
Command line and from IDE, restart, clean - nothing helps.

Reply:
I confirm that this issue reproduces after closing Android Studio, and running:
```
cd android; ./gradlew --stop; rm -rf .idea .gradle ~/.gradle
```

Adding this line into settings.gradle helps, although the toolchain is downloaded instead of being detected:
```diff
diff --git a/android/settings.gradle b/android/settings.gradle
index 46e6fb6146..0309398937 100644
--- a/android/settings.gradle
+++ b/android/settings.gradle
@@ -5,6 +5,11 @@ pluginManagement {
         gradlePluginPortal()
     }
 }
+
+plugins {
+    id 'org.gradle.toolchains.foojay-resolver-convention' version '0.8.0'
+}
+
 // See comment in buildscript block of app/build.gradle
 //dependencyResolutionManagement {
 //    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
```

Gradle team [says](https://youtrack.jetbrains.com/issue/KTIJ-24981/Gradle-8.-project-sync-fails-with-an-error-No-matching-toolchains-found-for-requested-specification-if-there-is-no-necessary-JDK) that it should be used when toolchain can not be resolved. But Java toolchain is always available/embedded in the Android Studio, and is usually installed from brew on Mac:

```
./gradlew -q javaToolchains
Building without Google Firebase Services
Version: 2024.11.16-3
VersionCode: 24111603
Building for [armeabi-v7a, arm64-v8a, x86_64] archs.
Create separate apks: false

 + Options
     | Auto-detection:     Enabled
     | Auto-download:      Enabled

 + Homebrew JDK 23.0.1
     | Location:           /opt/homebrew/Cellar/openjdk/23.0.1/libexec/openjdk.jdk/Contents/Home
     | Language Version:   23
     | Vendor:             Homebrew
     | Architecture:       aarch64
     | Is JDK:             true
     | Detected by:        MacOS java_home

 + JetBrains JDK 21.0.3+-79915917-b509.11
     | Location:           /Applications/Android Studio.app/Contents/jbr/Contents/Home
     | Language Version:   21
     | Vendor:             JetBrains
     | Architecture:       aarch64
     | Is JDK:             true
     | Detected by:        Current JVM
```

It looks like a bug somewhere in AGP or in gradle. As we manually override `targetCompatibility` and `sourceCompatibility`, it is likely should be safe to merge this.

Reply:
Looks like you manually installed some Java for Mac.

Recent gradle and AGP run on Java 21+, so installing old Java 17 is not a clean option. It will lead to the mix of different JDK versions used during the build by different components.

Asking contributors to install something else in the system instead of using Android Studio/default brew java out of the box may also complicate the entry point to the project.